kafka-client/src/main/java/dev/responsive/kafka/internal/stores/SegmentedOperations.java

```diff
@@ -0,0 +1,306 @@
+      final SubPartitioner partitioner,
+      final TopicPartition changelog,
+      final InternalProcessorContext context,
+      final ResponsiveStoreRegistry storeRegistry,
+      final ResponsiveStoreRegistration registration,
+      final ResponsiveRestoreListener restoreListener
+  ) {
+    this.params = params;
+    this.keySpec = keySpec;
+    this.table = table;
+    this.buffer = buffer;
+    this.partitioner = partitioner;
+    this.changelog = changelog;
+    this.context = context;
+    this.storeRegistry = storeRegistry;
+    this.registration = registration;
+    this.restoreListener = restoreListener;
+  }
+  
+  @Override
+  public void put(final Stamped<Bytes> windowedKey, final byte[] value) {
+    buffer.put(windowedKey, value, context.timestamp());
+  }
+
+  @Override
+  public void delete(final Stamped<Bytes> windowedKey) {
+    buffer.tombstone(windowedKey, context.timestamp());
+  }
+
+  @Override
+  public byte[] fetch(final Bytes key, final long windowStartTime) {
+    final Result<Stamped<Bytes>> localResult = buffer.get(new Stamped<>(key, windowStartTime));
+    if (localResult != null)  {
+      return localResult.isTombstone ? null : localResult.value;
+    }
+
+    return table.fetch(
+        partitioner.partition(changelog.partition(), key),
+        key,
+        windowStartTime
+    );
+  }
+
+  @Override
+  public WindowStoreIterator<byte[]> fetch(
+      final Bytes key,
+      final long timeFrom,
+      final long timeTo
+  ) {
+    final Stamped<Bytes> from = new Stamped<>(key, timeFrom);
+    final Stamped<Bytes> to = new Stamped<>(key, timeTo);
+
+    final int subPartition = partitioner.partition(changelog.partition(), key);
+    return Iterators.windowed(
+        new LocalRemoteKvIterator<>(
+            buffer.range(from, to),
+            table.fetch(subPartition, key, timeFrom, timeTo),
+            keySpec
+        )
+    );
+  }
+
+  @Override
+  public KeyValueIterator<Windowed<Bytes>, byte[]> fetch(
+      final Bytes keyFrom,
+      final Bytes keyTo,
+      final long timeFrom,
+      final long timeTo
+  ) {
+    throw new UnsupportedOperationException("Not yet implemented.");
+  }
+
+  @Override
+  public KeyValueIterator<Windowed<Bytes>, byte[]> fetchAll(
+      final long timeFrom,
+      final long timeTo
+  ) {
+    throw new UnsupportedOperationException("Not yet implemented");
+  }
+
+  @Override
+  public KeyValueIterator<Windowed<Bytes>, byte[]> all() {
+    throw new UnsupportedOperationException("Not yet implemented");
+  }
+
+  @Override
+  public WindowStoreIterator<byte[]> backwardFetch(
+      final Bytes key,
+      final long timeFrom,
+      final long timeTo
+  ) {
+    final Stamped<Bytes> from = new Stamped<>(key, timeFrom);
+    final Stamped<Bytes> to = new Stamped<>(key, timeTo);
+
+    final int subPartition = partitioner.partition(changelog.partition(), key);
+    return Iterators.windowed(
+        new LocalRemoteKvIterator<>(
```

Comment:
I think `LocalRemoteKVIterator` may have a bug in reverse fetching since it assumes keys are compared in ascending order - we should probably pass in the iteration order

Reply:
Good catch -- I'm just going to file a followup ticket for this for now and tackle it after the main window store PR. I added another new kind of iterator in the next PR for segmented range scans but was planning to leave the implementation of it for a followup. I'll tack this onto that patch

Reply:
FYI: https://www.notion.so/responsivedev/Iterators-for-range-scans-6aa2622b2b0b4b01a95eaabdebba347c?pvs=4